### ISSUE TITLE
Unable to use random chars in a string argument

### ISSUE THREAD

**Issue by CyborTronik** [bilal-fazlani/commanddotnet#268]
First of all thank you for this framework, is really easy to use. Good UX overall which often opensource projects have lack of.
----
I have declared a simple string argument for a command. All works fine unless I start to use non alpha numeric chars.
This doesn't work in prompt mode and as cli parameter.
Try out to put a value like: `sample: [yaml, array]`
For argument line I would expect something like `--param_name "sample: [yaml, array]"`
And in prompt mode just to put same text, but no way, chars like `:[,]` aren't accepted and no clear way to allow that happening

**Comment by CyborTronik** [comment]
I'm highly confused that a simple string cannot contain any char

**Comment by CyborTronik** [comment]
Overcome the issue with option:
`.UsePrompting(context => new CustomPrompter(context.Console))` where CustomPrompter is basically copy/paste `Prompter` class.
And highly disapointed about extensiability of this framework. Almost total violation of Open/Close principle (and few other moments form S.O.L.I.D.) forces you to copy and paste much of internals instead of extending them. Highly overused 'internal' keyword.

**Comment by CyborTronik** [comment]
This make it work in prompt mode. While the issue still persist for arguments. Unable to have a value like `"sample: [else]"` yet

**Comment by drewburlingame** [comment]
Hi @CyborTronik 

> First of all thank you for this framework, is really easy to use

You're welcome.  I'm glad you're liking it.

> highly disapointed about extensiability of this framework. Almost total violation of Open/Close principle

I think we may have different interpretations of the S.O.L.I.D.  In my understanding of it, the fact that you can extend the functionality of the framework by injected a new class is a class example of Open/Close principle.  Using interfaces instead of superclasses is generally considered a better way to do this.  [This post](https://stackify.com/solid-design-open-closed-principle/) covers that well.

And there are in fact multiple ways to extend this framework as seen under the `Extensibility` heading [in the documentation](https://commanddotnet.bilal-fazlani.com/) so I'm a bit confused.

> forces you to copy and paste much of internals instead of extending them. Highly overused internal keyword

This is intentional.  It allows the maintainers greater scope in what can be changed in refactorings.  If everything is public, then every change is a breaking change and the code base becomes difficult to change.  Even now, the code base has a lot of [Obsolete] tags and we're keeping [this page](https://commanddotnet.bilal-fazlani.com/ReleaseNotes/roadmap/) updated with everything that will change in the next major release.  Imagine how more cumbersome that would be if everything was public?  I would no longer want to contribute to this project.

If there is code that you'd like to be made public to make building apps easier, please make specific requests for those along with clear use cases how it will help.  In the case with Prompter, I'm guessing there are some behaviors you want and some you don't.  We would need to figure out a better design to expose that functionality publicly.

> is basically copy/paste Prompter class.

The [Promper](https://github.com/bilal-fazlani/commanddotnet/blob/master/CommandDotNet/Prompts/Prompter.cs) class is actually one of the public classes.  The core logic is implemented in private methods.  It sounds like there is a very specific way you wanted to be able to extend this class and the framework didn't support that.  The way you've described extending it though sounds like a standard approach though.

It sounds like with your approach you've fixed the bug you've mentioned here.  If that's the case, can you fork the repo and create a PR that would solve the problems you've encountered?

> Unable to have an argument value like "sample: [else]" yet

I'm not able to repro this issue.  We have [tests to cover special characters](https://github.com/bilal-fazlani/commanddotnet/blob/master/CommandDotNet.Tests/FeatureTests/Arguments/BasicParseTests.cs#L57)

* `"~!@#$%^&*()_= +[]\\{} |;':\",./<>?"`
* `"[some (parenthesis) {curly} and [bracketed] text]"`

and I just wrote these tests on a local branch and they pass

```c#
new AppRunner<App>().Verify(new Scenario
{
	When = { ArgsArray = new[] { "Do", "sample: [yaml, array]" } },
	Then =
	{
		AssertContext = ctx => ctx.ParamValuesShouldBe("sample: [yaml, array]")
	}
});
new AppRunner<App>().Verify(new Scenario
{
	When = { ArgsArray = new[] { "Options", "--opt1", "sample: [yaml, array]" } },
	Then =
	{
		AssertContext = ctx => ctx.ParamValuesShouldBe("sample: [yaml, array]", null)
	}
});
```

I'm using the `ArgsArray` so that I'm testing data as it would be passed in from the `Main(string[] args)`

I wonder if the problem you're having is caused by the way the shell is parsing the bug.

Some questions & suggestions to help troubleshoot.

* What OSs and shells have you tried this with and is the bug the same for all of them?
* Which version of this framework are you using?
* Have you tried using the `[parse:t]` directive to see how the arguments are being passed into the application?  Can you post the output here?
* If you can send a link to a repo with an app that has this issue with a full cli string to repro the problem and the shell used, that will be helpful for me to reproduce the issue.

**Comment by drewburlingame** [comment]
It would also be helpful to know what features are enabled.  For example, do you have directives enabled and does the first argument look like a directive?

You could post [Command Logger](https://commanddotnet.bilal-fazlani.com/Diagnostics/command-logger/) output with system info (excluding user & machine).

**Comment by drewburlingame** [comment]
I also tested this with the CommandDotNet.Example app as shown below and was able to enter the string you mentioned both as shell input and as prompt.  I'll definitely need more info to figure out what's going on for you there.

```bash
~/src/commanddotnet (master)
λ cmdnetex prompts echo

text (Text): sample: [yaml, array]
sample: [yaml, array]
~/src/commanddotnet (master)
λ cmdnetex prompts echo "sample: [yaml, array]"

sample: [yaml, array]
```

**Comment by drewburlingame** [comment]
@CyborTronik have you had a chance to look into this?  If I don't hear back in the next day I'll close this issue and you can reopen it when you've got more info to help troubleshoot.

**Comment by CyborTronik** [comment]
Sry, I was busy with my daily things.
The fix that I've did was by copy/paste and change the line https://github.com/bilal-fazlani/commanddotnet/blob/0d8994824a863703517a1f7bf5dded347235e78c/CommandDotNet/Prompts/Prompter.cs#L129
the `ConsoleKeys.PassThroughKeys` itslef won't allow any other characters than listed in it. 
Had no enough time to figure out where is the validation/tokenization issue for cli arguments yet

**Comment by drewburlingame** [comment]
I can't repro the problem you've listed with the example app.  Using the example app, I'm able to enter the text with brackets and I see it output correctly.  See the attached screenshot.

![image](https://user-images.githubusercontent.com/103345/80760929-95235300-8b31-11ea-8470-951ea4eb30c8.png)

I'll need the info as requested earlier:  os, .net version, CommandDotNet version and [parse:t] directive output.  If you can link to a repository with the bug, that would also be helpful.

**Comment by CyborTronik** [comment]
@drewburlingame this is because you are on totally other branch(v4/cancelation). while I'm using master (v3.x)

**Comment by drewburlingame** [comment]
@CyborTronik that branch didn't have any changes that would impact parsing.  But to remove all doubt...

![image](https://user-images.githubusercontent.com/103345/80807245-921d7680-8bb4-11ea-8642-3c9f934d04bc.png)

**Comment by CyborTronik** [comment]
I'll try few hours latter to make an unit test for it

**Comment by drewburlingame** [comment]
Great.  Also, please enable the [Parse directive](https://commanddotnet.bilal-fazlani.com/Diagnostics/parse-directive/) and run [parse:t].  Post the results along with the OS, .net version & CommandDotNet version.

**Comment by drewburlingame** [comment]
Those tests pass for me in Windows VS 2019 & Rider.

.Net version  = .NET Core 4.6.26614.01
OS version    = Microsoft Windows 10.0.18363

I'll look into running them in Ubuntu over the weekend.

**Comment by drewburlingame** [comment]
@CyborTronik what is the error message you get?

**Comment by CyborTronik** [comment]
They require to install dotnet 2
```Test run for /home/stanislav/workspace/commanddotnet/CommandDotNet.Tests/bin/Debug/netcoreapp2.0/CommandDotNet.Tests.dll(.NETCoreApp,Version=v2.0)
Microsoft (R) Test Execution Command Line Tool Version 16.5.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.
Testhost process exited with error: It was not possible to find any compatible framework version
The framework 'Microsoft.NETCore.App', version '2.0.0' was not found.
  - The following frameworks were found:
      3.1.3 at [/snap/dotnet-sdk/69/shared/Microsoft.NETCore.App]
You can resolve the problem by installing the specified framework and/or SDK.
The specified framework can be found at:
  - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=2.0.0&arch=x64&rid=ubuntu.18.04-x64
. Please check the diagnostic logs for more information.

Test Run Aborted.
```

**Comment by drewburlingame** [comment]
Oh, I thought you had errors for the tests you wrote.  This is a problem running the tests in general.  We're supporting 2.0.0 so you'd need to install that to run the tests.

**Comment by drewburlingame** [comment]
I setup an Ubuntu 19 image with dotnet-sdk-3.1 & Rider.  I had to change the target framework of CommandDotNet.Tests to netcoreapp3.1 to get Rider to find the tests.  Then I was able to run your tests and they passed for me.  I also built the example app and it worked as expected.  I can't repro this.   In your app, enable the Parse directive and use [parse:t] and paste the results here.

![image](https://user-images.githubusercontent.com/103345/80859874-d6b01d00-8c5b-11ea-9be3-4bec3f4f6ad2.png)
